compiler.md

```diff
@@ -1,155 +1,483 @@
+  from the architecture dependent backends
+- new language features are not useful by just being implemented
+- need to be fast (at least matching transpiled code), related optimizations are easier with
+  new pipeline
+- need to support debugging and be inspectable, this is achieved via better integration with
+  Chrome DevTools
+- new language features are easier optimized which makes them useable after much shorter time
+  after they are introduced to V8 (previously performance issues for new features prevented
+  their use in code that needed to run fast)
+- performance of ES6 features relative to the ES5 baseline operations per second tracked at [sixspeed](http://incaseofstairs.com/six-speed/)
+- at this point ES6 features are almost on par with ES5 versions of same code for most cases
+
+#### New Language Features Support And Transpilers
+
+[watch how to leverage babel optimally](https://youtu.be/HDuSEbLWyOY?t=15m5s)| [read deploying es2015 code](https://philipwalton.com/articles/deploying-es2015-code-in-production-today/)
+
+- using features directly, instead of transpiling, results in smaller code size [watch](https://youtu.be/HDuSEbLWyOY?t=13m)
+- additionally less parse time for untranspiled code and easier optimized
+- use [babel-preset-env](https://github.com/babel/babel/tree/master/packages/babel-preset-env) to specify browsers to target
+- therefore transpile es2015+ selectively
+
+### Resources
+
+- [Digging into the TurboFan JIT](https://v8project.blogspot.com/2015/07/digging-into-turbofan-jit.html)
+
+## Ignition Interpreter
+
+[watch](https://youtu.be/EdFDJANJJLs?t=13m16s) | [read](https://v8project.blogspot.com/2016/08/firing-up-ignition-interpreter.html)
+
+- uses TurboFan's low-level architecture-independent macro-assembly instructions to generate
+  bytecode handlers for each _opcode_
+- TurboFan compiles these instructions to target architecture including low-level instruction
+  selection and machine register allocation
+- bytecode passes through inline-optimization stages as it is generated
+  - common patterns replaced with faster sequences
+  - redundant operations removed
+  - minimize number of register transfers
+- this results in highly optimized and small interpreter code which can execute the bytecode instructions
+  and interact with rest of V8 VM in low overhead manner
+- Ignition Interpreter uses a [register machine](https://en.wikipedia.org/wiki/Register_machine)
+  with each bytecode specifying inputs and outputs as explicit register operands
+- holds its local state in _interpreter registers_
+  - some map to _real_ CPU registers
+  - others map to specific slots in native machine _stack memory_
+- last computed value of each bytecode is kept in special _accumulator_ register minimizing
+  load/store operations (from/to explicit registers)
+- current stack frame is identified by stack pointer
+- program counter points to currently executed instruction in the bytecode
```

Comment:
- each bytecode handler tail-calls into the next bytecode handler (indirectly threaded interpreter)

Reply:
Addressed in fe82604